code/modules/projectiles/pins.dm

```diff
@@ -0,0 +1,202 @@
+	gun.pin = null
+	gun = null
+	qdel(src)
+	return
+
+/obj/item/device/firing_pin/proc/pin_auth(mob/living/user)
+	return 1
+
+/obj/item/device/firing_pin/proc/auth_fail(mob/living/user)
+	user.show_message(fail_message, 1)
+	if(selfdestruct)
+		user.show_message("<span class='danger'>SELF-DESTRUCTING...</span><br>", 1)
+		to_chat(user, "<span class='userdanger'>[gun] explodes!</span>")
+		explosion(get_turf(gun), -1, 0, 2, 3)
+		if(gun)
+			qdel(gun)
+
+/*
+Pins Below.
+*/
+
+//only used in wizard staffs/wands.
+/obj/item/device/firing_pin/magic
+	name = "magic crystal shard"
+	desc = "A small enchanted shard which allows magical weapons to fire."
+	icon_state = "firing_pin_wizwoz"
+
+// Test pin, works only near firing range.
+/obj/item/device/firing_pin/test_range
+	name = "test-range firing pin"
+	desc = "This safety firing pin allows weapons to be fired within proximity to a firing range."
+	fail_message = "<span class='warning'>TEST RANGE CHECK FAILED.</span>"
+	pin_removeable = 1
+	origin_tech = list(TECH_MATERIAL = 2, TECH_COMBAT = 2)
+
+/obj/item/device/firing_pin/test_range/pin_auth(mob/living/user)
+	var/area/A = get_area(src)
+	if (A && (A.flags & FIRING_RANGE))
+		return 1
+	else
+		return 0
+
+// Implant pin, checks for implant
+/obj/item/device/firing_pin/implant
+	name = "implant-keyed firing pin"
+	desc = "This is a implant-locked firing pin which only authorizes users who are implanted with a certain device."
+	fail_message = "<span class='warning'>IMPLANT CHECK FAILED.</span>"
+	var/req_implant
+
+/obj/item/device/firing_pin/implant/pin_auth(mob/living/user)
+	if (locate(req_implant) in user)
+		return 1
+	else
+		return 0
+
+/obj/item/device/firing_pin/implant/loyalty
+	name = "loyalty firing pin"
+	desc = "This implant-locked firing pin authorizes the weapon for only loyalty-implanted users."
+	icon_state = "firing_pin_loyalty"
+	req_implant = /obj/item/weapon/implant/loyalty
+	
+// Honk pin, clown's joke item.
+// Can replace other pins. Replace a pin in cap's laser for extra fun! This is generally adminbus only unless someone thinks of a use for it.
+/obj/item/device/firing_pin/clown
+	name = "hilarious firing pin"
+	desc = "Advanced clowntech that can convert any firearm into a far more useful object."
+	color = "#FFFF00"
+	fail_message = "<span class='warning'>HONK!</span>"
+	force_replace = 1
+
+/obj/item/device/firing_pin/clown/pin_auth(mob/living/user)
+	playsound(src.loc, 'sound/items/bikehorn.ogg', 50, 1)
+	return 0
+
+// DNA-keyed pin.
+// When you want to keep your toys for youself.
+/obj/item/device/firing_pin/dna
+	name = "DNA-keyed firing pin"
+	desc = "This is a DNA-locked firing pin which only authorizes one user. Attempt to fire once to DNA-link."
+	icon_state = "firing_pin_dna"
+	fail_message = "<span class='warning'>DNA CHECK FAILED.</span>"
+	var/unique_enzymes = null
+
+/obj/item/device/firing_pin/dna/afterattack(atom/target, mob/user, proximity_flag)
+	..()
+	if(proximity_flag && iscarbon(target))
```

Comment:
Wouldnt that cause the DNA lock to be set every time you fire the gun ?
I think you need to check for `unique_enzymes == null` aswell

Reply:
it works fine in testing, I don't THINK it needs to be changed.